alpa/pipeline_parallel/cross_mesh_resharding.py

```diff
@@ -1179,3 +1491,433 @@ def _args_between(src_stage, dst_stage):
+        best_concurrent_selected_senders = []
+        for _ in range(n_rounds):
+            one_concurrent_works_ids, one_concurrent_selected_senders = \
+                self.find_one_random_concurrent_set_of_works(works_ids)
+            num = len(one_concurrent_works_ids)
+            if num < max_num:
+                continue
+
+            loads = list(self.loads)
+            for work_id, selected_sender in zip(
+                    one_concurrent_works_ids, one_concurrent_selected_senders):
+                loads[selected_sender] += self.works[work_id].duration
+
+            # here we could use different criterions
+            std = calc_std(loads)  # calc_max(loads)
+            # std = calc_std(
+            # [self.works[i].duration for i in range(one_concurrent_works_ids)]
+            # )
+
+            if num > max_num or (num == max_num and std < min_std):
+                max_num = num
+                min_std = std
+                best_concurrent_works_ids = one_concurrent_works_ids
+                best_concurrent_selected_senders = (
+                    one_concurrent_selected_senders)
+        return best_concurrent_works_ids, best_concurrent_selected_senders
+
+    def solve(self):
+        sol_assigned_sender_id = [None for _ in range(len(self.works))]
+        sol_order = []
+        while True:
+            available_works_ids = [
+                i for i in range(len(self.works)) if i not in sol_order
+            ]
+            best_concurrent_works_ids, best_concurrent_selected_senders = \
+                self.find_best_concurrent_set_of_works(available_works_ids)
+
+            for work_id, sender_id in zip(best_concurrent_works_ids,
+                                          best_concurrent_selected_senders):
+                sol_order.append(work_id)
+                sol_assigned_sender_id[work_id] = sender_id
+                self.loads[sender_id] += self.works[work_id].duration
+
+            if len(sol_order) == len(self.works):
+                break
+
+        assert None not in sol_assigned_sender_id
+        # debug
+        # print("order")
+        # print(sol_order)
+        # for i in sol_order:
+        #     print(self.works[i], sol_assigned_sender_id[i])
+
+        return sol_assigned_sender_id, sol_order
+
+
+class AbstractedLoadBalancingTaskSolverSearchAlgo(
+        AbstractedLoadBalancingTaskSolver):
+    """Implementation of load balance: use search algorithm with pruning"""
+
+    def __init__(self, n_workers, works):
+        super().__init__(n_workers, works)
+
+        self.sol_assigned_sender_id = [None for _ in range(len(self.works))]
+        self.sol_order = []
+        self.minimal_finish_time = None
+
+        self.cur_assigned_sender_id = [None for _ in range(len(self.works))]
+        self.cur_order = []
+
+        self.start_time = time.time()
+        self.search_time_threshold = 1
+
+    def evaluate_one_solution(self, assigned_sender_id, order):
```

Comment:
please add doc strings for different methods

Reply:
OK